app/domain/authentication/authn/login.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'command_class'
+
+module Authentication
+  module Authn
+    Login = C= CommandClass.new(
+      dependencies: {
+        role_cls:        ::Role,
+        credentials_cls: ::Credentials
+      },
+      inputs:       [:authenticator_input]
+    ) do
+
+      extend Forwardable
+      def_delegators :@authenticator_input, :account, :credentials, :username
+
+      def call
+        return nil unless validate_role_has_credentials
+        authenticate
+        api_key
+      end
+
+      private
+
+      def validate_role_has_credentials
+        role_credentials
+      end
+
+      def authenticate
+        @success = role_credentials.authenticate(credentials)
+      end
+
+      def api_key
+        @success ? role_credentials.api_key : nil
+      end
```

Comment:
what worried me in this solution is that you must call the `authenticate` method before you call `api_key` and it is not enforced. I'd rather have 
```
def api_key
  @success = role_credentials.authenticate(credentials)
  @success ? role_credentials.api_key : nil
end
```

and not call `authenticate` in the call function. you can change the method name to `api_key_if_authenticated`

Reply:
this is another option:
```
def api_key
  authenticate
  @success ? role_credentials.api_key : nil
end

def authenticate
  @success = role_credentials.authenticate(credentials)
end
```

it's similar to what you have but ensures that `authenticate` is called before we get the api_key.